src/Agent.Sdk/Util/TeeUtil.cs

```diff
@@ -0,0 +1,131 @@
+// Licensed under the MIT License.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Compression;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.Services.Agent.Util
+{
+    public static class TeeUtil
+    {
+        private static readonly string TeeTempDir = "tee_temp_dir";
+
+        private static readonly string TeeUrl = "https://vstsagenttools.blob.core.windows.net/tools/tee/14_135_0/TEE-CLC-14.135.0.zip";
+
+        public static void DownloadTeeIfAbsent(
+            string agentHomeDirectory,
+            string agentTempDirectory,
+            int providedDownloadRetryCount,
+            Action<string> debug,
+            CancellationToken cancellationToken
+        ) {
+            if (Directory.Exists(GetTeePath(agentHomeDirectory)))
+            {
+                return;
+            }
+
+            int downloadRetryCount = Math.Min(Math.Max(providedDownloadRetryCount, 3), 10);
+
+            for (int downloadAttempt = 1; downloadAttempt <= downloadRetryCount; downloadAttempt++)
+            {
+                try
+                {
+                    debug($"Trying to download and extract TEE. Attempt: {downloadAttempt}");
+                    DownloadAndExtractTee(agentHomeDirectory, agentTempDirectory, debug, cancellationToken);
+                    break;
+                }
+                catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
+                {
+                    debug("TEE download has been cancelled.");
+                    break;
+                }
+                catch (Exception ex) when (downloadAttempt != downloadRetryCount)
+                {
+                    debug($"Failed to download TEE. Error: {ex.ToString()}");
+                }
+            }
+        }
+
+        private static void DownloadAndExtractTee(
+            string agentHomeDirectory,
+            string agentTempDirectory,
+            Action<string> debug,
+            CancellationToken cancellationToken
+        ) {
+            string tempDirectory = Path.Combine(agentTempDirectory, TeeTempDir);
+            IOUtil.DeleteDirectory(tempDirectory, CancellationToken.None);
+            Directory.CreateDirectory(tempDirectory);
+
+            string zipPath = Path.Combine(tempDirectory, $"{Guid.NewGuid().ToString()}.zip");
+            DownloadTee(zipPath, debug, cancellationToken).GetAwaiter().GetResult();
+
+            debug($"Downloaded {zipPath}");
+
+            string extractedTeePath = Path.Combine(tempDirectory, $"{Guid.NewGuid().ToString()}");
+            ZipFile.ExtractToDirectory(zipPath, extractedTeePath);
+
+            debug($"Extracted {zipPath} to ${extractedTeePath}");
+
+            string extractedTeeDestinationPath = GetTeePath(agentHomeDirectory);
+            Directory.Move(Path.Combine(extractedTeePath, "TEE-CLC-14.135.0"), extractedTeeDestinationPath);
```

Comment:
Nit: seems to be a duplication of plugin name - could it make sense to move to one constant?

Reply:
Moved it to a static constant